working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,421 @@
+language communities is that declarative-only metaprogramming ends up being too
+limited.
+
+The Scheme community has gone through a number of rules-based macro systems over
+decades as each proved to be too limiting. Meanwhile Common Lisp let Lispers
+define macros imperatively. This sometimes makes it a challenge to implement
+macros *correctly*, but has always been sufficiently powerful.
+
+Experience from many UI templating languages (which are surprisingly similar to
+macro metaprogramming in many ways) is that most declarative ones gradually
+accumulate imperative features or end up replaced by those that do.
+
+In the Dart ecosystem, in the absence of metaprogramming in the language, users
+often resort to code generators. Some of those are quite sophisticated and are
+written in imperative Dart code.
+
+Rust has a nice [declarative, rule-based macro system][rust_declarative_macros].
+But, because that rule-based system is limited, Rust also has a second
+[imperative procedural macro system][rust_procedural_macros] for when you need
+more power. In Dart, we would prefer to only have one metaprogramming system, so
+it should be one that is powerful enough to cover all known use cases.
+
+## Challenges
+
+Having said that, allowing the whole Dart language to run at compile-time
+exposes a number of really difficult problems we will have to solve in 2021:
+
+### Scoping and hygiene
+
+A macro may introduce new symbols into an existing scope (where that function is
+used). How do we deal with conflicts here? Would the generated code exist in its
+own scope and not have the ability to leak any new variables into the
+surrounding scope?
+
+Generated code may also need to access variables in at least 3 different scopes —
+the scope where the macro is applied, the scope of the generated code (may be
+same as the former), and the scope where the macro is defined. How does it
+distinguish between these?
+
+### Ordering and staging
+
+In what order are macros run? If they can both generate new APIs and introspect
+on the program, then ordering can be user-visible. Consider a macro that adds a
+field for memoization and another that adds a method to serialize all fields. If
+both are applied to the same class, does the memoized field get serialized?
+
+Do macros run before type checking, after, or interleaved? What is the static
+type context in which a macro runs?
+
+There are additional considerations if we allow recursive macros, and cycles
+between macro defining libraries.
+
+### Usability
+
+Macros can easily lead to incomprehensible programs if they become overly
+magical, or are a complete black box to the user. In order to demystify macros
+you should:
+
+* Be able to easily navigate to their implementation code.
+* Be able to visualize in some way the code they generated in your program, at
+  development time.
+* Be able to debug generated code (step through it, etc).
```

Comment:
probably also need to be able to opt to not-debug into it unless you really want to.

Reply:
We don't have that for build_runner generated code today - I added that as a possibility but I don't see it as a requirement (nobody has asked for it in build_runner for instance).

Reply:
It probably depends on how invasive the uses of static metaprogramming are. If they are all use cases that code generators could also solve I agree that always having the code always be debuggable is fine. If the use cases are more invasive then accidentally stepping into the generated code could be quite annoying. For example, stepping into some synthetic code for every get of a `disposable` field would be a bit annoying just as stepping into the synthetic getter for a `late` field would be annoying.